Source/WebCore/accessibility/AXObjectCache.cpp

```diff
@@ -4060,6 +4060,18 @@ void AXObjectCache::updateRelationsIfNeeded()
         // Collect all possible origins, i.e., elements with non-empty relation attributes.
         for (const auto& attribute : relationAttributes()) {
             auto& idsString = element.attributeWithoutSynchronization(attribute);
+            if (idsString.isEmpty()) {
```

Comment:
This isn't right. Even when the id string isn't empty, we should be preferring the element reference over ID string.
Consider a case an element moves from being the same tree to being in an "outer" tree.

Reply:
Good point, even if we just simply change the `id` of the element that is being referenced, we'd be losing that relationship  from the a11y POV. So indeed we have to trust the element reference over the ID information. I'll change that and add tests.

Simple example that breaks due to this (we'll be getting null instead of the `foo/new-foo` element:
```html
<div id="target">target</div>
<div id="foo">foo</div>
<script>
  target.ariaDescribedByElements = [foo];
  foo.id = "new-foo";
  console.log(accessibilityController.accessibleElementById("target").ariaDescribedByElementAtIndex(0));
</script>
```

Reply:
And thinking about the `id` there might be another issue from the spec POV, what's going on if we have something like this?

```html
<div id="target">target</div>
<div id="foo1">foo1</div>
<div id="foo2">foo2</div>

<script>
  let myFoo1 = document.getElementById("foo1");
  let myFoo2 = document.getElementById("foo2");
  
  target.ariaDescribedByElements = [myFoo1];
  
  myFoo1.id = "foo2";
  myFoo2.id = "foo1";
  
  console.log("aria-describedby attribute: " + target.getAttribute("aria-describedby"));
  console.log("ariaDescribedByElements: " + target.ariaDescribedByElements[0].id);
</script>
```

Should we trust the attribute or the element reference? I guess I need to fill another HTML issue about this. :/

Reply:
JFTR filled https://github.com/whatwg/html/issues/8307

Reply:
I've changed this on the implementation so it follows element reference and not the ID. That's also what Chromium does.